src/main/java/io/debezium/connector/cassandra/SchemaHolder.java

```diff
@@ -5,154 +5,305 @@
+            }
+            catch (Exception e) {
+                LOGGER.warn("Error happened while updating the keyspace {} in Schema instance", current.getName(), e);
+            }
         }
 
-        public Schema keySchema() {
-            return keySchema;
+        @Override
+        public void onKeyspaceRemoved(final KeyspaceMetadata keyspaceMetadata) {
+            try {
+                Schema.instance.clearKeyspaceMetadata(org.apache.cassandra.schema.KeyspaceMetadata.create(
+                        keyspaceMetadata.getName(),
+                        KeyspaceParams.create(keyspaceMetadata.isDurableWrites(),
+                                keyspaceMetadata.getReplication())));
+                LOGGER.info("Removed keyspace [{}] from Schema Instance.", keyspaceMetadata.asCQLQuery());
+            }
+            catch (Exception e) {
+                LOGGER.warn("Error happened while removing the keyspace {} from Schema instance", keyspaceMetadata.getName(), e);
+            }
         }
 
-        public Schema valueSchema() {
-            return valueSchema;
+        @Override
+        public void onTableAdded(final TableMetadata tableMetadata) {
+            if (tableMetadata.getOptions().isCDC()) {
+                addOrUpdateTableSchema(new KeyspaceTable(tableMetadata),
+                        new KeyValueSchema(kafkaTopicPrefix, tableMetadata, sourceInfoStructMaker));
+            }
+            try {
+                LOGGER.debug("Table {}.{} detected to be added!", tableMetadata.getKeyspace().getName(), tableMetadata.getName());
+                final CFMetaData rawCFMetaData = CFMetaData.compile(tableMetadata.asCQLQuery(), tableMetadata.getKeyspace().getName());
+                // we need to copy because CFMetaData.compile will generate new cfId which would not match id of old metadata
+                final CFMetaData newCFMetaData = rawCFMetaData.copy(tableMetadata.getId());
+                Keyspace.open(newCFMetaData.ksName).initCf(newCFMetaData, false);
+                final org.apache.cassandra.schema.KeyspaceMetadata current = Schema.instance.getKSMetaData(newCFMetaData.ksName);
+                if (current == null) {
+                    LOGGER.warn("Keyspace {} doesn't exist", newCFMetaData.ksName);
+                    return;
+                }
+                if (current.tables.get(tableMetadata.getName()).isPresent()) {
+                    LOGGER.debug("Table {}.{} is already added!", tableMetadata.getKeyspace(), tableMetadata.getName());
+                    return;
+                }
+                final java.util.function.Function<org.apache.cassandra.schema.KeyspaceMetadata, org.apache.cassandra.schema.KeyspaceMetadata> transformationFunction = ks -> ks
+                        .withSwapped(ks.tables.with(newCFMetaData));
+                org.apache.cassandra.schema.KeyspaceMetadata transformed = transformationFunction.apply(current);
```

Comment:
Why going through the `Function` indirection here?

Reply:
> Why going through the `Function` indirection here?

I'm not sure about this, just copying the codes from SchemaProcessor landed in PR#26.

Reply:
Ok, will push a commit for simplifying this and merge. Thx!